Mathlib/CategoryTheory/Yoneda.lean

```diff
@@ -504,4 +506,201 @@ lemma isIso_of_yoneda_map_bijective {X Y : C} (f : X ⟶ Y)
+
+@[simp]
+theorem coyonedaEvaluation_map_down (P Q : C × (C ⥤ Type v₁)) (α : P ⟶ Q)
+    (x : (coyonedaEvaluation C).obj P) :
+    ((coyonedaEvaluation C).map α x).down = α.2.app Q.1 (P.2.map α.1 x.down) :=
+  rfl
+
+/-- The "Coyoneda pairing" functor, which sends `X : C` and `F : C ⥤ Type`
+to `coyoneda.rightOp.obj X ⟶ F`, functorially in both `X` and `F`.
+-/
+def coyonedaPairing : C × (C ⥤ Type v₁) ⥤ Type max u₁ v₁ :=
+  Functor.prod coyoneda.rightOp (𝟭 (C ⥤ Type v₁)) ⋙ Functor.hom (C ⥤ Type v₁)
+
+-- Porting note: we need to provide this `@[ext]` lemma separately,
+-- as `ext` will not look through the definition.
+-- See https://github.com/leanprover-community/mathlib4/issues/5229
+@[ext]
+lemma coyonedaPairingExt {x y : (coyonedaPairing C).obj X} (w : ∀ Y, x.app Y = y.app Y) : x = y :=
+  NatTrans.ext _ _ (funext w)
+
+@[simp]
+theorem coyonedaPairing_map (P Q : C × (C ⥤ Type v₁)) (α : P ⟶ Q) (β : (coyonedaPairing C).obj P) :
+    (coyonedaPairing C).map α β = coyoneda.map α.1.op ≫ β ≫ α.2 :=
+  rfl
+
+variable {C} in
+/-- A bijection `(coyoneda.obj X ⋙ uliftFunctor ⟶ F) ≃ F.obj (unop X)` which is a variant
+of `coyonedaEquiv` with heterogeneous universes. -/
+def coyonedaCompUliftFunctorEquiv (F : C ⥤ Type max v₁ w) (X : Cᵒᵖ) :
+    (coyoneda.obj X ⋙ uliftFunctor.{w} ⟶ F) ≃ F.obj X.unop where
+  toFun φ := φ.app X.unop (ULift.up (𝟙 _))
+  invFun f :=
+    { app := fun Y x => F.map (ULift.down x) f }
+  left_inv φ := by
+    ext Y f
+    dsimp
+    rw [← FunctorToTypes.naturality]
+    dsimp
+    rw [Category.id_comp]
+    rfl
+  right_inv f := by aesop_cat
+
+/-- The Coyoneda lemma asserts that the Coyoneda pairing
+`(X : C, F : C ⥤ Type) ↦ (coyoneda.obj X ⟶ F)`
+is naturally isomorphic to the evaluation `(X, F) ↦ F.obj X`.
+
+See <https://stacks.math.columbia.edu/tag/001P>.
+-/
+def coyonedaLemma : coyonedaPairing C ≅ coyonedaEvaluation C where
+  hom :=
+    { app := fun F x => ULift.up ((x.app F.1) (𝟙 F.1))
+      naturality := by
+        intro X Y f
+        ext
+        simp [coyonedaEvaluation, ← FunctorToTypes.naturality] }
+  inv :=
+    { app := fun F x =>
+        { app := fun X a => (F.2.map a) x.down }
+      naturality := by
+        intro X Y f
+        ext
+        simp [yoneda, ← FunctorToTypes.naturality] }
+  hom_inv_id := by
+    ext
+    simp [← FunctorToTypes.naturality]
+  inv_hom_id := by
+    ext
+    simp [ULift.up_down]
+
+variable {C}
+
+/-- The isomorphism between `coyoneda.obj X ⟶ F` and `F.obj (unop X)`
+(we need to insert a `ULift` to get the universes right!)
+given by the Coyoneda lemma.
+-/
+@[simps!]
+def coyonedaSections (X : Cᵒᵖ) (F : C ⥤ Type v₁) :
+    (coyoneda.obj X ⟶ F) ≅ ULift.{u₁} (F.obj X.unop) :=
+  (coyonedaLemma C).app (X.unop, F)
+
+/-- We have a type-level equivalence between natural transformations from the coyoneda embedding
+and elements of `F.obj X.unop`, without any universe switching.
+-/
+def coyonedaEquiv {X : Cᵒᵖ} {F : C ⥤ Type v₁} : (coyoneda.obj X ⟶ F) ≃ F.obj X.unop :=
```

Comment:
I think that this equivalence should be defined first, and everything else should be deduced from it. (The same refactor should be done also for `yonedaEquiv`.)

Reply:
Are you saying that `coyonedaLemma` should become:
```lean
def coyonedaLemma : coyonedaPairing C ≅ coyonedaEvaluation C where
  hom := {
      app := fun (X, F) x ↦ ULift.up (coyonedaEquiv x)
      naturality := by
        intro (X, F) (Y, G) f
        ext a
        simp [coyonedaEvaluation, ← FunctorToTypes.naturality, coyonedaEquiv]
    }
  inv := {
      app := fun (X, F) x ↦ coyonedaEquiv.symm x.down
      naturality := by
        intro X Y f
        ext a b c
        simp [← FunctorToTypes.naturality, coyonedaEquiv]
    }
  hom_inv_id := by
    ext
    simp
  inv_hom_id := by
    ext
    simp
    rfl
```
where `coyonedaEquiv` is moved up (and changed to take `X : C` instead of `X : Cᵒᵖ`, since this has better definitional properties (gets rid of some `erw`s).

Reply:
Yes, mostly, though a more direct definition can be obtained as follows:
```lean
def coyonedaEquiv {X : Cᵒᵖ} {F : C ⥤ Type v₁} : (coyoneda.obj X ⟶ F) ≃ F.obj X.unop where
  toFun φ := φ.app _ (𝟙 _)
  invFun := sorry
  left_inv := sorry
  right_inv := sorry

def coyonedaLemma : coyonedaPairing C ≅ coyonedaEvaluation C :=
  NatIso.ofComponents (fun X => Equiv.toIso (coyonedaEquiv.trans Equiv.ulift.symm)) sorry
```

Reply:
Please check the latest commit to see if this is what you wanted. In the proof of `coyonedaLemma` appear two `erw`s which I don't understand. I tried to debug with `pp.explicit` but did not find why the lemma does not apply with `rw` alone.

Reply:
I am not worried about the two `erw`. It seems it is related to issues with coercions to functions (which cause a lot of similar problems, and I am qualified to know how this could be fixed...).

Reply:
If you are happy with the `co`-part, I'll adjust the non-`co` part accordingly.

Reply:
Apart from the question of the fate of `coyonedaSections*`, it looks good to me! (I am sorry that while proceeding to this PR, you did not expect that I would trick you into doing a refactor!)